Migrate chat app from Gitter to a better alternative (#20)
enarx/rfcs: concepts/Enarx_Public_Chat_App_Zulip/000005-Enarx_Public_Chat_App_Zulip.md

| Sign-up ease                                | Easy (Github login)                                                                                                                                           | Easy (Github, Google, email login) -- worth noting email confirmations usually take a long time to arrive                                                                             |
| Direct linking to specific messages in chat | No                                                                                                                                                            | Yes                                                                                                                                                                                   |
| File uploads/image sharing                  | No                                                                                                                                                            | Yes                                                                                                                                                                                   |

> | Bridging to other networks                  | Barebones                                                                                                                                                     | Officially for [Matrix](https://zulip.com/integrations/doc/matrix) and [IRC](https://zulip.com/integrations/doc/irc), but functionally broken (only syncs a stream and not a channel) |

Reviewer: ```suggestion
| Bridging to other networks                  | Barebones                                                                                                                                                     | Officially for [Matrix](https://zulip.com/integrations/doc/matrix) and [IRC](https://zulip.com/integrations/doc/irc), but functionally broken (only syncs a single topic in a stream and not the whole stream) |
```

## Drawbacks

There are two major drawbacks to Zulip we should strongly consider before a move.

**Community**

The community we’ve begun building up on Gitter is easily the biggest reason we have to stay. If we move chat clients, we’d be asking them to juggle around yet another chat app, and their existing lines of communication would end up unsupported by us. This is not ideal, and a big con to moving.

The counter-argument to this is that if a change is going to happen, it’s better that it happens sooner rather than later. If we truly feel that Gitter won’t meet our needs (especially as we scale), a change now will be far less painful than when we’re a lot larger.

**Public chat visibility**

The biggest thing Gitter does significantly _better_ than Zulip is the ability to read messages when logged out. If someone new to the project is idly curious what's going on in chat, they can just look at our chat log in Gitter -- no login required.

Zulip, on the other hand, requires a login, which presents a barrier to people being able to see the discussions going on about Enarx. An idle observer would need to create an account, which they might not want to do.

> There is a potential workaround present, however, with [`zulip-archive`](https://github.com/zulip/zulip-archive). This is a tool, runnable as a Github Action, that archives our chat logs and publishes them on a publicly-accessible Github Pages site. For Enarx, this may manifest itself as a sub-link of `enarx.github.io` -- for example, we may be able to point people to `enarx.github.io/zulip-archive` for a publicly-accessible chat record.

Reviewer: Two small things.

I think we should avoid using the enarx.github.io domain and stick to enarx.io, which is better in terms of "branding".

We could also make the name of the chat more generic, for future proofing. So `chat.enarx.io` as a canonical chat URL and maybe `chat.enarx.io/archive`.

```suggestion
There is a potential workaround present, however, with [`zulip-archive`](https://github.com/zulip/zulip-archive). This is a tool, runnable as a Github Action, that archives our chat logs and publishes them on a publicly-accessible Github Pages site. For Enarx, this may manifest itself as a sub-link of `enarx.io` -- for example, we may be able to point people to `chat.enarx..io/archive` for a publicly-accessible chat record.
```

- App support (especially mobile)
    - Gitter’s mobile app is, to put it kindly, not very good. It doesn’t support notifications at all, it is slow, and it doesn’t register read messages well. Additionally, Gitter themselves do not officially recommend their own apps, and [say](https://gitter.im/apps) they may be deprecated in the future. Zulip’s mobile apps work well, do support notifications, and generally support a larger subset of features.
    - Both have desktop Linux apps available, though both are also available in browser.
    - Both have Windows and Mac desktop clients.
- Nice-to-have features
    - Zulip supports file uploads, where Gitter does not. This can be helpful if we need to share a screenshot, for example.
    - Zulip allows users to link to a unique message in the chat.
    - Zulip appears to have much more robust integration support, including Github integration. Gitter’s Github integration is out of date and does not work.

In terms of alternatives: Following Enarx's design principles of open development, we strongly value FOSS solutions, of which there are, unfortunately, few. Beyond Zulip, there's:

> - Gitter, which we've already discussed

Reviewer: ```suggestion
- [Gitter](https://gitter.im), which we've already discussed
```

> - Matrix, which has some nice security/privacy features, but otherwise looks high-maintenance and not necessarily as feature-complete as Zulip.

Reviewer: ```suggestion
- [Matrix](https://matrix.org), which has some nice security/privacy features, but otherwise looks high-maintenance and not necessarily as feature-complete as Zulip.
```

Reviewer: I'm not sure what you mean by high-maintenance. The 3 options right now for using matrix as a chat system are:
1. public servers: create accounts on a public server (ie: matrix.org)
2. self-hosted: run your own server and create accounts there
3. paid hosting: this is [what Mozilla is doing](https://chat.mozilla.org/#/welcome), using modular.im

1 and 3 require no effort, but 3 requires money. 2 requires effort.

Threaded conversations and custom emojis are not available in any Matrix client right now, and "communities" (groups of channels/chats) are in need of more work to make them nice to use (which is planned) otherwise, i think everything else is on par with Zulip, or superior (good bridging to multiple other networks, end-to-end encryption).

I don't want to make a big case for Matrix, as i think the end-user experience is still a little less polished than Zulip's (even though that will change in the near future).

But it is decentralised, which means that the conversation can keep going when a server goes down (although users on that server are obviously disconnected).

Author: All fair points. I've amended the section to be more accurate (in truth, I have little experience with Matrix proper, so your feedback is valuable).

- IRC, which is a non-starter.

## Prior art

We have used Gitter for a while, and have had annoyances with it for a while, as previously discussed. Beyond that, Enarx has not used any other public chat apps.

## Unresolved questions

We still have no reference point for how reliable Zulip is. This, unfortunately, is something we can only learn with experience.

## Future Possibilities

> One particularly exciting possibility is Zulip's robust integration potential. Zulip has a powerful [bot](https://zulipchat.com/api/running-bots) framework, which could potentially open up interesting automation opportunities with Github and elsewhere.

Reviewer: This is both true and already frustrating. I expect Zulip's integration with other non-chat platforms is better than it is with chat platforms, where, as mentioned in the RFC document, you can only sync a Zulip **topic**  to another platform (which has very limited usefulness, given many topics will be created). But i expect the integration with other platforms that are complementary will be better.

Reviewer 2: @axelsimon I'm not sure what else could be done. If you support topics (Zulip) and you want to sync to a platform that doesn't support topics (IRC), the best possible integration is to lock IRC to a single topic.

Reviewer: That's a good point Nathaniel. I think i'd just at least sync everything, adding the topic to the message text, rather than only sync one topic, but at least a switch to chose which of these two workarounds to use would be nice. I suspect locking to a single topic can also have its use in that you can have a topic for, say, monitoring, and have that replicated in other places.

Anyway, that's taking us on a bit of a tangent.
